interview-questions.md

```diff
@@ -7,60 +7,61 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 
 
 2. What is a block in Ruby?
 
-  Your answer:
+  Your answer: Blocks in Ruby is code that is enclosed in a do end statement or curly braces; {}. Blocks are functions that can take arguments inside of two pipes. 
 
-  Researched answer:
+  Researched answer: A block consists of chunks of code that is always enclosed in braces. Blocks are invoked from a function with the same name as the block. Blocks can be invoked by using the yield statement. Yield statements can also be used to pass parameters to block code. Blocks in Ruby can also be declared to be run as the file is being loaded and after the program has finished executing with the "begin" and "end" blocks.
 
 
 
 3. Ruby has an implicit return. What does that mean?
 
-  Your answer:
+  Your answer: In Ruby an implicit return is where inside of functions we do not have to explicitly write return. The return is implied as the last part of the code block. 
 
-  Researched answer:
+  Researched answer: In Ruby with implicit return, the value of the last line of a method is automatically returned without using the keyword return. Ruby's methods will return the result of the last evaluated expression. For example, a ruby method that "return a + b" can be simply written as "a + b " in the code block without the return keyword and will return the sum of a + b just the same.
 
 
 
 4. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
+  Your answer: Object-oriented programming is the style of programming that focuses on classes and objects. OOP is about using blueprints/classes to make different types of objects, methods, etc to create a program. Functional programming is where the focus is on applying and creating functions. 
 
-  Researched answer:
+  Researched answer: The four principles of object-oriented programming are encapsulation, abstraction, inheritance, and polymorphism. Encapsulation is when each object keeps its state private inside of a class, meaning where other objects don't have direct access to this state unless explicitly allowed. Abstraction is like an extension of encapsulation where the idea is to only reveal operations relevant for other objects, and hiding internal implementation details. An example would be having public methods that any class can call without "knowing" how they work, a lot like how we use our cell phones but have no idea how they work under the hood. Inheritance is a way to reuse common logic and extract unique logic into seperate classes. A child class can be derived from a parent class, by doing so the child class "inherits" all of the properties and methods of the parent class and can also have it's own unique values. Polymorphism is where child classes can implment their own versions of the a parent classes methods. 
```

Comment:
This is a similar answer to one I used in an actual interview! Highly recommend hanging on to this!